### PR TITLE
[mlir][Vector] Replace `vector.shuffle` with `vector.interleave` in vector narrow type emulation

### PR DESCRIPTION
**Pull request by dcaballe**
This PR replaces the generation of `vector.shuffle` with `vector.interleave` in the i4 conversions in vector narrow type emulation. The multi dimensional semantics of `vector.interleave` allow us to enable these conversion emulations also for multi dimensional vectors.

### CODE REVIEW

**Review by hanhanW — APPROVED** [review]
It looks better, thanks!

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 1]
```diff
@@ -879,8 +881,7 @@ static Value rewriteI4ToI8SignedExt(PatternRewriter &rewriter, Location loc,
     interleaveMaskValues.push_back(i + (vecDimSize / 2));
   }
```
Comment (MacDue) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:882]: Can the `interleaveMaskValues` (and the loop generate it) be removed now? It looks unused.

Also `// 3. Interleave low and high i8 elements using a shuffle.` is now outdated :)

File: mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp [hunk 2]
```diff
@@ -724,9 +724,8 @@ BitCastRewriter::BitCastRewriter(VectorType sourceVectorType,
 static LogicalResult commonConversionPrecondition(PatternRewriter &rewriter,
                                                   VectorType preconditionType,
                                                   Operation *op) {
-  if (!preconditionType || preconditionType.getRank() != 1 ||
-      preconditionType.isScalable())
-    return rewriter.notifyMatchFailure(op, "scalable or >1-D vector");
+  if (!preconditionType || preconditionType.isScalable())
+    return rewriter.notifyMatchFailure(op, "scalable vector");
```
Comment (MacDue) [inline mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:728]: Is there anything stopping allowing this for scalable vectors too now?

Comment (dcaballe) [reply mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:728]: I would say broader validation to make sure the backend is ok with the `i4` pieces but I currently don't have a large workload that compiles using scalable vectors.

**Review by MacDue — APPROVED** [review]
LGTM!

**Pull request merged**